File: src/api/mcprofile.go
package api

import (
	"encoding/base64"
	"encoding/json"
	"fmt"

	"github.com/devproje/project-api/src/utils"
)

type extractor struct {
	Username string `json:"name"`
	UUID     string `json:"id"`
}

type EncodedMojangAPI struct {
	Username   string `json:"name"`
	UUID       string `json:"id"`
	Properties []struct {
		Name  string `json:"name"`
		Value string `json:"value"`
	} `json:"properties"`
}

type MojangAPI struct {
	UUID     string `json:"profileId"`
	Username string `json:"profileName"`
	Textures struct {
		Skin struct {
			Url string `json:"url"`
		} `json:"SKIN"`
	} `json:"textures"`
}

func RefactorUUID(uuid string) string {
	var dash = func(str string, index int) string {
		return fmt.Sprintf("%s-%s", str[:index], str[index:])
	}
	i := 8
	return dash(dash(dash(dash(uuid, i), i+5), i+5*2), i+5*3)
}

func getUUID(username string) (*extractor, int) {
	url := fmt.Sprintf("https://api.mojang.com/users/profiles/minecraft/%s", username)
	data, err := utils.GET[extractor](url)
	if err != nil {
		return nil, 204
	}

	return data, 200
}

func decodeProperty(encoded EncodedMojangAPI) (*MojangAPI, error) {
	bytes, err := base64.StdEncoding.DecodeString(encoded.Properties[0].Value)
	if err != nil {
		return nil, err
	}

	var decoded MojangAPI
	err = json.Unmarshal(bytes, &decoded)
	if err != nil {
		return nil, err
	}

	return &decoded, nil
}

func GetMCProfile(username string) (*MojangAPI, int) {
	var uuid, status = getUUID(username)
	if status != 200 {
		return nil, status
	}

	url := fmt.Sprintf("https://sessionserver.mojang.com/session/minecraft/profile/%s", uuid.UUID)
	e, err := utils.GET[EncodedMojangAPI](url)
	if err != nil {
		return nil, 204
	}

	data, err := decodeProperty(*e)
	if err != nil {
		return nil, 500
	}

	return data, status
}

File: src/main/kotlin/net/projecttl/api/MCProfile.kt
package net.projecttl.api

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import net.projecttl.model.*
import net.projecttl.utils.unwrapQuote
import java.lang.String.format
import java.net.URL
import java.nio.charset.Charset
import java.util.Base64
import java.util.UUID

class MCProfile(private val username: String) {
    private val uuid: UUID
        get() {
            val url = URL(format("https://api.mojang.com/users/profiles/minecraft/%s", username))
            val obj = Json.parseToJsonElement(url.readText(Charset.forName("UTF-8")))
            val rawUUID = unwrapQuote(obj.jsonObject["id"]!!)

            return UUID.fromString(refactorUUID(rawUUID))
        }

    private fun refactorUUID(str: String): String {
        fun String.parse(range: IntRange, end: Boolean = false): String {
            var ps = ""
            for (i in range) {
                ps += this[i]
            }

            if (!end) {
                ps += "-"
            }

            return ps
        }
        var parsed = ""
        parsed += str.parse(0..7)
        parsed += str.parse(8..11)
        parsed += str.parse(12..15)
        parsed += str.parse(16..19)
        parsed += str.parse(20..31, true)

        return parsed
    }

    private fun decode(encoded: String): String {
        return Base64.getDecoder().decode(encoded).decodeToString()
    }

    fun getProfile(): MojangAPI {
        val start = System.currentTimeMillis()
        val url = URL(format("https://sessionserver.mojang.com/session/minecraft/profile/%s", uuid))
        val rawJson = Json.parseToJsonElement(url.readText(Charset.forName("UTF-8")))
        val encoded = unwrapQuote(rawJson.jsonObject["properties"]!!.jsonArray[0].jsonObject["value"]!!)
        val decoded = Json.parseToJsonElement(decode(encoded))
        val cape = try {
            Cape(unwrapQuote(decoded.jsonObject["textures"]!!.jsonObject["CAPE"]!!.jsonObject["url"]!!))
        } catch (ex: Exception) {
            null
        }

        val resTime = System.currentTimeMillis() - start
        return MojangAPI(
            status = 200,
            unique_id = unwrapQuote(decoded.jsonObject["profileId"]!!),
            username = unwrapQuote(decoded.jsonObject["profileName"]!!),
            textures = Textures(
                skin = Skin(
                    url = unwrapQuote(decoded.jsonObject["textures"]!!.jsonObject["SKIN"]!!.jsonObject["url"]!!)
                ),
                cape = cape
            ),
            res_time = "${resTime}ms"
        )
    }
}